clients/src/main/java/org/apache/kafka/clients/KafkaClient.java

```diff
@@ -86,7 +86,17 @@
     List<ClientResponse> poll(long timeout, long now);
 
     /**
+     * Diconnects the connection to a particular node, if there is one.
+     * Any pending ClientRequests for this connection will receive disconnections.
+     *
+     * @param nodeId The id of the node
+     */
+    void disconnect(String nodeId);
```

Comment:
@junrao One difference is that the producer only handles time outs before it sends a request. The reason why the AdminClient has to do a little more is that it has per request timeouts. I think the right solution is to add per request timeouts to `NetworkClient`, but that was a bigger change and we postponed that for a subsequent release.

Having said that, it's a good question if we could just invoke `Call.fail` by retrieving it from `callsInFlight`. I had a look at the code and it's unclear if it makes things simpler. It certainly keeps `NetworkClient` simpler, but `AdminClient` becomes a little more complex. Do you think it's worth trying to do this now? One option is to go with what's in the PR and then implement per request timeouts in NetworkClient.

What do you think?